### PR TITLE
non_exhaustive for provider crates

### PR DESCRIPTION
**Pull request by Manishearth**
Progress on #170, #383

### CODE REVIEW

File: provider/adapters/src/lib.rs [hunk 1]
```diff
@@ -16,7 +16,9 @@
         clippy::indexing_slicing,
         clippy::unwrap_used,
         clippy::expect_used,
-        clippy::panic
+        clippy::panic,
+        clippy::exhaustive_structs,
+        clippy::exhaustive_enums
```
Comment (robertbastian) [inline provider/adapters/src/lib.rs:21]: Nit: can this have a trailing comma?

File: provider/core/src/data_provider.rs [hunk 2]
```diff
@@ -27,6 +27,7 @@ use icu_locid::LanguageIdentifier;
 pub struct DataRequestMetadata;
 
 #[derive(Default, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::exhaustive_structs)] // this type is stable
```
Comment (robertbastian) [inline provider/core/src/data_provider.rs:30]: Nit: "this type is stable", "stable type", "stable"?

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: provider/adapters/src/lib.rs (hunk 1 above)
Comment (Manishearth) [reply provider/adapters/src/lib.rs:21]: i believe it can't according to rustfmt

File: provider/core/src/data_provider.rs (hunk 2 above)
Comment (Manishearth) [reply provider/core/src/data_provider.rs:30]: yeah we're a bit inconsistent across the codebase about this

**Review by robertbastian — DISMISSED** [review]

File: provider/testdata/src/metadata.rs [hunk 3]
```diff
@@ -34,6 +35,7 @@ impl From<serde_json::Error> for Error {
 }
 
 #[derive(Debug, Deserialize)]
+#[allow(clippy::exhaustive_structs)] // this type is stable
```
Comment (sffc) [inline provider/testdata/src/metadata.rs:38]: Nit: I think not, because we may want to add more keys to it

**Review by sffc — CHANGES_REQUESTED** [review]

**Review by sffc — DISMISSED** [review]

**Review by zbraniecki — APPROVED** [review]
lgtm

**Pull request merged**